kubernetes/resource_kubernetes_certificate_signing_request.go

```diff
@@ -0,0 +1,196 @@
+				Type: "Approved",
+				Reason: "TerraformAutoApprove",
+				Message: "This CSR was approved by Terraform auto_approve.",
+			}
+			pendingCSR.Status.Certificate = []byte{}
+			pendingCSR.Status.Conditions = append(pendingCSR.Status.Conditions, approval)
+			_, updateErr := conn.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(pendingCSR)
+			return updateErr
+		})
+	if retryErr != nil {
+		panic(fmt.Errorf("CSR auto-approve update failed: %v", retryErr))
+	}
+	fmt.Println("CSR auto-approve update succeeded")
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Printf("[DEBUG] Waiting for certificate to be issued")
+	stateConf := &resource.StateChangeConf{
+			Target:  []string{"Approved"},
+			Pending: []string{""},
+			Timeout: d.Timeout(schema.TimeoutCreate),
+			Refresh: func() (interface{}, string, error) {
+				out, refreshErr := conn.CertificatesV1beta1().CertificateSigningRequests().Get(metadata.Name, metav1.GetOptions{})
+				if refreshErr != nil {
+					log.Printf("[ERROR] Received error: %#v", refreshErr)
+					return out, "Error", err
+				}
+				statusConditions := ""
+				emptyStatus := v1beta1.CertificateSigningRequestStatus{}
+
+				if !reflect.DeepEqual(emptyStatus, out.Status) {
+					statusConditions = fmt.Sprintf("%v", out.Status.Conditions[0].Type)
+					if statusConditions == "Approved" && string(out.Status.Certificate) == "" {
+						fmt.Println("CSR status: 'Approved', but certificate has not been issued yet. Retrying...")
+						statusConditions = ""
+					}
+				}
+				log.Printf("[DEBUG] CertificateSigningRequest %s status received: %#v", out.Name, statusConditions)
+				return out, statusConditions, nil
+			},
+		}
+		_, err = stateConf.WaitForState()
+		if err != nil {
+			return err
+		}
+		log.Printf("[INFO] Certificate issued for request: %s", metadata.Name)
+
+		finalRead, err := conn.CertificatesV1beta1().CertificateSigningRequests().Get(metadata.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// Delete the remote CSR object, since the certificate is stored in terraform state.
+		err = conn.CertificatesV1beta1().CertificateSigningRequests().Delete(metadata.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			log.Printf("[ERROR] Unable to clean up after certificate creation: %#v", err)
+			return err
+		}
+
+		d.SetId(metadata.Name)
+		d.Set("certificate", string(finalRead.Status.Certificate))
+
+	return resourceKubernetesCertificateSigningRequestRead(d, meta)
+}
+
+// resourceKubernetesCertificateSigningRequestRead does not return any data, because Read functions exist to
+// sync the local state with the remote state. Since this data is local-only, there is nothing to read.
+func resourceKubernetesCertificateSigningRequestRead(d *schema.ResourceData, meta interface{}) error {
+	_ = d
+	_ = meta
+	return nil
+}
+
+func resourceKubernetesCertificateSigningRequestDelete(d *schema.ResourceData, meta interface{}) error {
+	_ = meta
```

Comment:
I put this `_ = meta` here because otherwise the parameter `meta` is unused. But the SDK requires the parameter even if you're not using it.

Reply:
Couldn't you make `_` the name of the parameter in the function signature, instead of doing this assignment?

Reply:
I don't think that the underscoring thing is necessary at all for function arguments.
If I'm not mistaken, Go doesn't account the function arguments as "unused variables" when they aren't actually user.

Did a quick test here: https://play.golang.com/p/G6NaZdO6x0O

Reply:
Yeah, the code runs without `_`, but IDEs will flag it as an unused variable, depending on the IDE's linter. Mine underlines unused variables. This is what I see: https://i.imgur.com/BcxGmQd.png

But I've renamed the unused variables to `_` like John suggested and I think that looks better.

Reply:
I looked up some information about why Go allows code to compile when there is an unused parameter. The purpose is to allow better readability in the function when you have optional parameters. I guess it's also required if you have more than one optional parameter too.

But since this isn't a parameter that will ever be used in these functions, I think it makes sense to leave it as `_`. (Reference https://stackoverflow.com/a/22549439/9231732 )